### PR TITLE
feat(auth): add auth to armory cloud

### PR DESCRIPTION
**Pull request by ivanphdz**

### CODE REVIEW

File: cmd/deploy/start.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+		Example: deployStartExample,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return start(cmd,deployOptions, args)
+		},
+	}
+	return cmd
+}
+
+func start(cmd *cobra.Command, options *cmd.RootOptions, args []string) error {
+	logrus.Info("Not implemented")
```
Comment (fieldju) [inline cmd/deploy/start.go:30]: FatelF
```suggestion
	logrus.FatelF("Not implemented") // this will cause the CLI to exit 1
```

Comment (fieldju) [reply cmd/deploy/start.go:30]: // this will cause the CLI to exit 1

File: pkg/auth/auth.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		return "", err
+	}
+	if exists {
+		currentCreds, err := LoadCredentials(dirname + "/.armory/credentials")
+		if err != nil {
+			return "", err
+		}
+
+		expiresAt, err := time.Parse("2006-01-02 15:04:05", currentCreds.ExpiresAt)
```
Comment (danielpeach) [inline pkg/auth/auth.go:53]: I think you want `time.Parse(time.RFC3339, currentCreds.ExpiresAt)` which will parse the date as an ISO formatted string.

File: cmd/deploy/start.go (hunk 1 above)
Comment (DanJohnstonArmory) [reply cmd/deploy/start.go:30]: @fieldju - is that supposed to be FatalF?

File: cmd/version/version.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package version
+
+import (
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var Version = "development"
```
[the remark this answers is not in the document]
Comment (ivanphdz) [reply cmd/version/version.go:8]: I make them const but version is a variable which is set on the build of the project.
https://github.com/armory/armory-cli/blob/master/Makefile#L25

File: pkg/auth/auth.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+		secret:         clientSecret,
+		source:         source,
+		tokenIssuerUrl: "__tokenIssuerUrl__",
+		audience:       "__audience__",
+		verify:         true,
+	}
+}
+
+func (a *Auth) GetToken() (string, error) {
+	dirname, err := os.UserHomeDir()
```
Comment (fieldju) [inline pkg/auth/auth.go:39]: What happens when the user doesn't have a home dir?
Like in an alpine docker container with just the CLI?

File: pkg/auth/auth.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+		if err != nil {
+			return "", err
+		}
+
+		expiresAt, err := time.Parse(time.RFC3339, currentCreds.ExpiresAt)
+		if err != nil {
+			return "", err
+		}
+
+		if time.Now().Before(expiresAt) && (a.clientId == "" || a.clientId == currentCreds.ClientId) {
```
Comment (fieldju) [inline pkg/auth/auth.go:58]: You need a buffer here, unless you are adding a buffer when setting the `expiresAt` prop originally.

Also what happens when credentials expire in the middle of a long command like watching a 1-2 hour production rollout.

File: pkg/auth/auth.go (hunk 4 above)
Comment (ivanphdz) [reply pkg/auth/auth.go:43]: It depends on the OS, it uses the default env variable of that OS, it returns an error if the env doesn't exists, 
if you don't set an user on alpine, alpine uses root, and set as home the root dir

File: pkg/auth/credentials.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	return &Credentials{
+		Audience:  audience,
+		Source:    source,
+		ClientId:  clientId,
+		ExpiresAt: expiresAt,
+		Token:     token,
+	}
+}
+
+func (c *Credentials) WriteCredentials(fileLocation string) error {
```
Comment (th3morg) [inline pkg/auth/credentials.go:26]: One thing that would be nice here is if we did a pretty print to format the json. You should be able to use:
`json.MarshalIndent(c, "", "    ")` which would use 4 spaces for the indentation.

File: pkg/auth/auth.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
+package auth
```
Comment (th3morg) [inline pkg/auth/auth.go:1]: I'd really love to see unit tests in the project, especially for the auth class. I think you (Ivan) and Gaby are working on them 👍

**Pull request merged**